44805 Updates to the Connecting to an Image Registry topic (#256)
replicatedhq/replicated-docs: docs/vendor/packaging-private-images.md

# Connecting to an Image Registry

When packaging and delivering an application with Replicated, you can include private
images for the application without distributing registry credentials
to your customer.

The customer license file can grant revokable image pull access to private images,

> whether the images are stored in the Replicated private registry or a supported

Reviewer: Should we link these to the headings? e.g.  [Replicated Private Registry](#push-images-to-the-replicated-private-registry)

Author: Yeah good call

external registry.

:::note
The app manager supports image tags for applications in all use cases.

> The app manager supports image digests only for online (Internet-connected) installation where the app manager can pull all images from the Replicated private registry, a public external registry, or from a private external registry through proxy access.

Author: This note about image tags vs digest was mixed in at the bottom of the old topic on kots.io.

I wasn't sure how this information might or might not factor into this topic. Is it helpful when you are deciding how to version images before doing any of these steps to connect to a registry?

Reviewer: this is a pretty niche topic, the note could probably land in a troubleshooting section at the end

Reviewer: reading again, this should definitely be near the bottom, in a troubleshooting or "other stuff". also could probably use some examples. e.g.

**okay**
- Image Tag `docker.io/postgres:10.2`
- Image Digest `docker.io/postgres@sha256dadf78ae715fea7ef3feafea0fe8afea5fe33`

**not okay**

- Image Tag w Digest `docker.io/postgres:10.2@sha256dadf78ae715fea7ef3feafea0fe8afea5fe33`

I also am not sure what is technically correct here, the second one under `okay` might also be `not okay` -- @AmberAlston do you or another PM know the answer here? Or can get it from an eng team?

Reviewer 2: Not that it should dictate the new location, but fyi that this content used to be at the bottom of the page on the old docs site. 

The example above is not technically correct, as is. 
* Image tag = works
* Image digest = only works in some use cases (this note content)
* image tag with digest = may still be broken (https://app.shortcut.com/replicated/story/28008/admin-console-using-a-docker-image-with-both-a-tag-and-a-digest-fails-when-checking-for-updates-to-pull-the-release). If fixed, it is anticipated to have the same use case limitations as digest note above.

:::

## About Using an External Registry

> If your application images are available in a private, accessible image registry, such as Docker Hub, quay.io, Amazon Elastic Container Registry (ECR), Google Container Registry (GCR), or Artifactory, then the customer licenses for your application can grant proxy, or _pull-through_, access to the assignee without exposing registry credentials to the customer.

Author: "the assignee" --> can we change this to "the app manager"?

Reviewer: "authenticating mirror" is a term that lands well with prospects, I'll have a full read through and see where that might fit in

> If your application images are available in a private, accessible image registry, such as Docker Hub, quay.io, Amazon Elastic Container Registry (ECR), Google Container Registry (GCR), or Artifactory, then the customer licenses for your application can grant proxy, or _pull-through_, access to the assignee without exposing registry credentials to the customer.

Reviewer: ```suggestion
If your application images are available in a private image registry exposed to the internet such as Docker Hub, quay.io, Amazon Elastic Container Registry (ECR), Google Container Registry (GCR), or Artifactory, then the customer licenses for your application can grant proxy, or _pull-through_, access to the assignee without exposing registry credentials to the customer.
```

Reviewer 3: As a general rule, we should remove possessive pronouns in favor of the passive voice. IE "If Vendor application images" instead of "If your application images" ... and "then the customer licenses for an application" instead of "then the customer licenses for your application"

Author: @tdensmore we actually go the other way for tech doc, where everything should be active voice. We also prefer the second person "you". Both active voice and sticking with the second person help improve clarity. For more info, you can check out our style guide: https://wiki.replicated.io/doc/grammar-and-formatting-28QRVfWVgb

With proxy access, the app manager can use the Replicated registry proxy service to pull private images from an external registry. Instances of your application can then pull private images from the proxy service at `proxy.replicated.com` during deployment.

> Allowing the app manager to have proxy access to an external registry is useful and recommended because it prevents you from having to modify the process you use to build and push application images in order to deploy your application with Replicated. It also allows you to revoke a customer’s ability to pull private images, such as when their trial license expires.

Reviewer 3: Connecting App Manager to a private Vendor registry via a proxy is useful and recommended because it a Vendor from having to modify the process used to build and push application images. It also allows Vendors to easily revoke a customer’s ability to pull private images (such as when a trial license expires), without having to manage image access through separate identity or auth systems.

The following diagram demonstrates how the registry proxy service pulls images from your external registry, and how deployed instances of your application pull images from the proxy service:

![Registry proxy service workflow diagram](/images/private-registry-diagram.png)

[View a larger version of this image](/images/private-registry-diagram-large.png)

For more information about how the app manager uses the registry proxy service, see [How the App Manager Accesses Private Images](#how-the-app-manager-accesses-private-images) below.

## Configure Access to an External Registry

You can provide the credentials for an external registry in the vendor portal to
grant the app manager proxy access to the private application images in the
registry.

All applications in your vendor portal Team have access to the external registry
that you add. This means that you can use the images in the external registry across
multiple apps in the Team.

To configure access to your private images in an external registry:

1. Log in to the [vendor portal](https://vendor.replicated.com) and go to the Images page.
1. Click **Add external registry**.

>    ![Add External Registry dialog in the vendor portal](/images/add-external-registry.png)

Reviewer: I think this screenshot is way out of date, can we get an updated product shot

Author: Yeah good catch, I had just updated this screenshot, but it looks like some of the apps in our vendor portal team have the wrong (out-dated) dialog? 🤷‍♀️ I found an app that has the correct dialog, the one which has that diagram in it.

1. Complete the fields in the dialog:
<table>
  <tr>
    <th width="30%">Field</th>
    <th width="70%">Instructions</th>
  </tr>
  <tr>
    <td>Name</td>
    <td>Enter a name for the registry.</td>
  </tr>
  <tr>
    <td>Endpoint</td>
    <td>Enter the endpoint, such as quay.io, index.docker.io, or gcr.io.</td>
  </tr>
  <tr>
    <td>Username and Password</td>
    <td>Provide the username and password to Replicated for an account that has pull access.<br/><br/>Replicated stores your username and password encrypted and securely. Your credentials and the encryption key do not leave Replicated servers.</td>
  </tr>
  <tr>
    <td>Email Address</td>

>     <td>Enter the email address associated with the registry.</td>

Reviewer: its generally deprecated and unused, i'm not sure why we still have it on the form. https://docs.docker.com/engine/deprecated/#-e-and---email-flags-on-docker-login

>     <td>Enter the email address associated with the registry.</td>

Author: Will remove this--related to the screenshot issue above, for some reason the app in our vendor portal team that I used to take an updated screenshot of this dialog was out of date. The current dialog of course does not have "Email address".

  </tr>
</table>

## Push Images to the Replicated Private Registry

> You can host the private images for your application on the Replicated private registry.

Author: Is there any more info we can add about why someone would want to use the replicated private registry? to help them decide if they want to use this feature

Author: If you don't already have images in another registry. Almost everyone has an external registry

Reviewer 3: I would say something like this: 
"Instead of using the Vendor Portal to proxy images from a private registry, the Vendor Portal can also act as a registry itself. This can be useful for Vendors that prefer to tag and push custom containers to the Vendor Portal for a quick proof of concept, or due to internal registry restrictions. "

Reviewer 3: (I have seen some Vendors that are in the middle of a transition to containers and their internal registry address was so unstable through the PoC that they decided to use the Vendor Portal as a registry.)

> For more information about building, tagging and pushing docker images, see the

Reviewer 3: I think Docker is always capitalized. Otherwise we should say container runtime.

[Docker CLI documentation](https://docs.docker.com/engine/reference/commandline/cli/).

To push images to the Replicated private registry:

1. Do one of the following to connect with the `registry.replicated.com` container registry:
   * **Log in with a service account or team token**: Use `docker login registry.replicated.com` with a Replicated vendor portal service account or team token as the password. You can use any string as the username. For more information, see [Service accounts](../reference/replicated-cli-tokens#service-accounts) and [Team tokens](../reference/replicated-cli-tokens#team-tokens) in _Using Vendor API tokens_.
   * **Log in with a user token**: Use `docker login registry.replicated.com` with your vendor portal email as the username and a vendor portal user token as the password. For more information, see [User tokens](../reference/replicated-cli-tokens#user-tokens) in _Using Vendor API tokens_.

>    * **Log in with your credentials**: Use `docker login registry.replicated.com` with your vendor portal email and password as the credentials.

Reviewer: ```suggestion
   * **Log in with your credentials**: Use `docker login registry.replicated.com` with your vendor portal email and password as the credentials. **Note**: this method is still allowed for backwards compatibility. Using a User Token is recommended instead.
```

1. Log in to the [vendor portal](https://vendor.replicated.com) and go to the Images page.
1. Click **Add external registry**.

   ![Add External Registry dialog in the vendor portal](/images/add-external-registry.png)

1. Click the **Other Registry** or **DockerHub** tab.
1. Complete the fields in the dialog:
<table>
  <tr>
    <th width="30%">Field</th>
    <th width="70%">Instructions</th>
  </tr>
  <tr>
    <td>Endpoint</td>
    <td>Enter the endpoint, such as quay.io, index.docker.io, or gcr.io.</td>
  </tr>
  <tr>
    <td>Username and Password</td>
    <td>Provide the username and password to Replicated for an account that has pull access.<br/><br/>Replicated stores your username and password encrypted and securely. Your credentials and the encryption key do not leave Replicated servers.</td>
  </tr>
</table>

## Push Images to the Replicated Private Registry

You can host the private images for your application on the Replicated private registry. Hosting your images on the Replicated private registry is useful if you do not already have your images in an existing private registry. It is also useful for testing purposes.

> For more information about building, tagging and pushing Docker images, see the

Reviewer 4: You could add a serial comma here if you want to. Not sure how you feel about them. I'm not a big fan of them, but it has been standard in my last few jobs.

Author: Oh good catch. I do love me a good serial comma

registry proxy service.

The app manager uses Kustomize to patch the `midstream/kustomization.yaml` file to change the image name during deployment to reference the proxy service.

For example, a PodSpec for a Deployment references a private image hosted at `quay.io/my-org/api:v1.0.1`:

```yaml
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: example
  spec:
    template:
      spec:
        containers:
          - name: api
            image: quay.io/my-org/api:v1.0.1
```

When this application is deployed, the app manager detects that it cannot access
the image at quay.io. So, it creates a patch in the `midstream/kustomization.yaml`
file that changes the image name in all manifest files for the application:

```yaml
  apiVersion: kustomize.config.k8s.io/v1beta1
  bases:
  - ../../base
  images:
  - name: quay.io/my-org/api:v1.0.1
    newName: proxy.replicated.com/proxy/my-kots-app/quay.io/my-org/api
```

This causes the container runtime in the cluster to use the proxy service to pull the images,
using the license information provided to the app manager for authentication.

### Accessing the Proxy Service with an Image Pull Secret

> During installation, the app manager automatically creates an `imagePullSecret`

Reviewer 4: Do we want to define "it"? It can be clearer for the user to define the noun. Also, if you can delete "in order" if you want to.

that is based on the customer license. It uses this secret to authenticate in order to
pull private images from `proxy.replicated.com`.

The app manager does not patch the image location URL for images hosted on the Replicated private registry

> at `registry.replicated.com`. However, it adds the same `imagePullSecret` to

Author: That's a great point--I went ahead and replaced these because it's a pretty detailed/complex chunk of content, and it (that is, repeating the noun) doesn't hurt 😉 

(I figure for less detailed or complex topics, "it" is probably fine. Though I like the idea of being sure to consider repeating the noun to add clarity)

PodSpecs that reference images in the Replicated private registry.

:::note
When deploying Pods to namespaces other than the app manager application namespace, you must add the namespace to the `additionalNamespaces` attribute of the Application custom resource manifest.

> This ensures the app manager can provision the the `imagePullSecret` in the namespace to allow the Pod to pull the image.

Reviewer 4: There's an extra "the" in the second sentence of this note.

Author: ty!

  $ docker push registry.replicated.com/APPLICATION_SLUG/TARGET_IMAGE_NAME:TAG
  ```

## How the App Manager Accesses Private Images

The app manager uses Kustomize to change the `name` of private images in the PodSpec

> during application deployment from the external registry domain to the Replicated

Author: These are my own words--want to confirm this idea of changing the "domain" is accurate

Reviewer 3: `change the location of the image registry` instead of `name` ?

proxy service domain.
It then creates an `imagePullSecret` to access the images through the proxy service.

### Patching the Image Name with Kustomize

When the app manager is installing an application, it attempts to load image manifests
using the image reference from the PodSpec. If the app manager receives a 401 response,
it assumes that this is a private image that must be proxied through the
registry proxy service.

The app manager uses Kustomize to patch the `midstream/kustomization.yaml` file to change the image name during deployment to reference the proxy service.

> For example, a Deployment and PodSpec reference a private image hosted at `quay.io/my-org/api:v1.0.1`:

Reviewer: ```suggestion
For example, a Deployment's PodSpec may reference a private image hosted at `quay.io/my-org/api:v1.0.1`:
```

```yaml
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: example
  spec:
    template:
      spec:
        containers:
          - name: api
            image: quay.io/my-org/api:v1.0.1
```

When this application is deployed, the app manager detects that it cannot access
the image at quay.io. So, it creates a patch in the `midstream/kustomization.yaml`

> file that changes the image name everywhere it appears:

Author: "everywhere it appears" --> where all might it appear? does this go beyond the podspec?

Reviewer: ```suggestion
file that changes the image name in all application YAML files:
```

```yaml
  apiVersion: kustomize.config.k8s.io/v1beta1
  bases:
  - ../../base
  images:
  - name: quay.io/my-org/api:v1.0.1
    newName: proxy.replicated.com/proxy/my-kots-app/quay.io/my-org/api
```

> This allows the app manager to use the proxy service to pull the images.

Author: Nice

### Accessing the Proxy Service with an Image Pull Secret

During installation, the app manager automatically creates an `imagePullSecret`
that is based on the customer license. It uses this secret to authenticate in order to
pull private images from `proxy.replicated.com`.

> The app manager does not overwrite images hosted on the Replicated private registry

Author: "the app manager does not overwrite images from the Replicated private registry"

This language came from the old topic. I'm wondering if this is accurate...it never overwrites images, correct? Just the image name in order to reference the proxy service?

Sounds like in the case of the Replicated private registry, there's no need to go through the proxy service?

Reviewer: ah good callout 
```suggestion
The app manager does not patch URLs for images hosted on the Replicated private registry
```
